protocol/x/clob/keeper/process_operations.go

```diff
@@ -248,6 +214,218 @@ func (k Keeper) PersistMatchToState(
 	return nil
 }
 
+// PersistOrderRemovalToState takes in an OrderRemoval, statefully validates it according to
+// RemovalReason, and writes the removal to state.
+func (k Keeper) PersistOrderRemovalToState(
+	ctx sdk.Context,
+	orderRemoval *types.OrderRemoval,
+) error {
+	orderIdToRemove := orderRemoval.GetOrderId()
+	orderIdToRemove.MustBeStatefulOrder()
+
+	// Order removals are always for stateful orders that must exist.
+	orderToRemove, err := k.FetchOrderFromOrderId(ctx, orderIdToRemove, make(map[types.OrderId]types.Order, 0))
+	if err != nil {
+		return err
+	}
+
+	// Statefully validate that the removal reason is valid.
+	switch removalReason := orderRemoval.RemovalReason; removalReason {
+	case types.OrderRemoval_REMOVAL_REASON_UNDERCOLLATERALIZED:
+		// For the collateralization check, use the remaining amount of the order that is resting on the book.
+		remainingAmount, hasRemainingAmount := k.MemClob.GetOrderRemainingAmount(ctx, orderToRemove)
```

Comment:
cleaner approach is to move `MemClob.GetOrderRemainingAmount` to the keeper. `GetOrderRemainingAmount` already calls keeper method anyways, so the logic to get remaining amount can also live on the keeper side.

Overall, we should try to avoid calling memclob methods if possible, since DeliverTx does not and should not modify memclob. Any modification should be delayed to prepare check state

Reply:
I agree, I can update this in a different PR though. I started to make some changes but it started to blow up a bit and involves fixing a bunch of mocks in tests

Reply:
https://linear.app/dydx/issue/CLOB-876/move-getorderremainingamount-into-clob-keeper